daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala

```diff
@@ -4,73 +4,142 @@
-  override def toString: String = "Error(" + msg + ")"
+import com.daml.lf.data.Ref
+import com.daml.lf.transaction.{GlobalKey, GlobalKeyWithMaintainers}
+import com.daml.lf.value.Value
+
+sealed abstract class Error {
+  def subError: Error.SubError
+  final def msg: String = subError.msg
+  final def detailMsg: String = msg
 }
 
 object Error {
 
-  def withError[A](e: Error)(optional: Option[A]): Either[Error, A] = {
-    optional.fold[Either[Error, A]](Left(e))(v => Right(v))
+  sealed abstract class SubError extends RuntimeException with scala.util.control.NoStackTrace {
+    // msg is intended to be a single line message
+    def msg: String
+    // details for debugging should be included here
+    def detailMsg: String
+    def toError: Error
   }
 
-  /** small conversion from option to either error with specific error */
-  implicit class optionError[A](o: Option[A]) {
-    def errorIfEmpty(e: Error) = withError(e)(o)
+  // Error happening during Package loading
+  final case class Package(subError: Package.SubError) extends Error
+
+  object Package {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Package = Package(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
+    final case class Generic(override val msg: String, override val detailMsg: String)
+        extends SubError
+    object Generic {
+      def apply(msg: String): Generic = new Generic(msg, msg)
+    }
+
+    final case class Validation(validationError: validation.ValidationError) extends SubError {
+      def msg: String = validationError.pretty
+    }
+
   }
 
-  def apply(description: String) = new Error {
-    val msg = description
-    override def detailMsg = msg
+  // Error happening during command/transaction preprocessing
+  final case class Preprocessing(subError: Preprocessing.SubError) extends Error
+
+  object Preprocessing {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Preprocessing = Preprocessing(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
+    final case class Generic(override val msg: String, override val detailMsg: String)
+        extends SubError
+    object Generic {
+      def apply(msg: String): Generic = new Generic(msg, msg)
+    }
+
+    final case class Lookup(lookupError: language.LookupError) extends SubError {
+      def msg: String = lookupError.pretty
+    }
+
+    private[engine] object MissingPackage {
+      def unapply(error: Lookup): Option[Ref.PackageId] =
+        error.lookupError match {
+          case language.LookupError.Package(packageId) => Some(packageId)
+          case _ => None
+        }
+    }
   }
 
-  def apply(description: String, details: String) = new Error {
-    val msg = description
-    val detailMsg = details
+  // Error happening during interpretation
+  final case class Interpretation(subError: Interpretation.SubError) extends Error
```

Comment:
I don’t understand the weird `SubError` stuff. How is this better than (simplified)
```
sealed abstract class SError
sealed abstract class InterpretationError extends SError
case class ContractNotFound(cid: ContractId) extends InterpretationError
```